webidls/Browser.webidl

```diff
@@ -0,0 +1,115 @@
+
+  // Will fail if Browser.session != pipeline.session.
+  // Useful with prerendering pipelines
+  Promise<Pipeline> createPrerenderingPipeline(LoadData loadData);
+  Promise<HistoryEntry> navigateToPipeline(Pipeline pipeline, optional Pipeline opener);
+
+
+  // FIXME: do we want to give access to autopurge and maxLivePipeline?
+  // Purging means killing the pipeline. The entry stays intact
+  readonly attribute boolean doPipelinesAutopurge;
+  // Maximum number of live pipeline before and after the current entry
+  Promise<void> enableAutoPurgePipelines(unsigned long before, unsigned long after);
+  Promise<void> disableAutoPurgePipelines();
+
+  // Popup blocker, tracking content blocker, mixed content blocker,
+  // and safari-like content blocker. See ContentBlockers.webidl
+  // FIXME: not cool. We can't have multiple custom blockers.
+  readonly attribute Sequence<ContentBlockerDescription> contentBlockers;
+  Promise<void> setContentBlockers(Sequence<ContentBlockerDescription>);
+
+  // FIXME: not sure focus should be handled here
+  readonly attribute boolean isFocused;
+
+  // FIXME: can we do without prefs?
+  readonly attribute Object prefs;
+  Promise<void> setPrefs(Object prefs); // use to set user-agent for example
+
+  Promise<boolean/*prevent default canceled*/> dispatchKeyEvent(KeyEvent event);
+}
+
+interface BrowserHandler {
+
+  // A new document is displayed. Usually after the user
+  // clicked on a link and once the new document has been
+  // created (pipeline is not pending anymore). Also happens
+  // when user or page goes back/forward.
+  void onActiveEntryChanged();
+
+  void onFocusChanged();
+  
+  // window.onClose() is called
+  Cancelable onDestroyRequestedFromContent();
+
+  void onDestroy();
+
+}
+
+// EXPERIMENTAL AND TEMPORARY
+
+// Below interfaces are not optimized for performance but for ease of
+// implementation.
+// 
+// The goal it to be able to experiment with non-linear history. Entries are
+// stored in an array. When a user goes back in the history and then navigates,
+// the previous forward history is dropped. We want to be able to experiment
+// with saving that previous forward history, moving from an array structure to
+// a tree structure. This goes against the web specifications, and might
+// involve intrusive changes in Servo.
+//
+// The following interfaces are a temporary solution that doesn't require much
+// work on Servo's side, and at the same time make it possible to build a tree
+// structure at the API consumer level. There are some drawbacks.  Pipeline's
+// are dropped, only the LoadData is saved, and the consumer has to maintain an
+// alternate history structure.
+
+partial interface Browser {
+  // Will drop the back and forward entries and replace with new ones.
+  // At that point, all pipelines are killed. Only the current one stays
+  // alive.
+  Promise<void> replaceEntriesButCurrent(FrozenArray<LoadData> pastLoadData, FrozenArray<LoadData> futureLoadData);
```

Comment:
If you'll go with a suggestion of adding `historyEntries`  to the constructor this becomes obsolete, as you will be able to achieve this by just creating new `Browser` with desired entries instead.